File: apps/mobile/lib/crdt/orSet.ts
/**
 * Observed-Remove Set (OR-Set) CRDT
 * A set that handles concurrent adds and removes correctly
 * Each element is tagged with unique tags that are tracked separately
 */

import { VectorClock, VectorClockUtils } from './vectorClock';

export interface ORSetTag {
  id: string;
  nodeId: string;
  vectorClock: VectorClock;
}

export interface ORSetElement<T> {
  value: T;
  tags: ORSetTag[];
}

export class ORSet<T> {
  private elements: Map<string, ORSetElement<T>>;

  constructor() {
    this.elements = new Map();
  }

  /**
   * Generate a unique tag for an element
   */
  private generateTag(nodeId: string): ORSetTag {
    return {
      id: `${nodeId}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
      nodeId,
      vectorClock: VectorClockUtils.create(nodeId, 1),
    };
  }

  /**
   * Generate a unique key for an element value
   */
  private generateKey(value: T): string {
    return JSON.stringify(value);
  }

  /**
   * Add an element to the set
   */
  add(value: T, nodeId: string): void {
    const key = this.generateKey(value);
    const tag = this.generateTag(nodeId);
    
    const existing = this.elements.get(key);
    
    if (existing) {
      // Add new tag to existing element
      existing.tags.push(tag);
    } else {
      // Create new element with tag
      this.elements.set(key, {
        value,
        tags: [tag],
      });
    }
  }

  /**
   * Remove an element from the set
   */
  remove(value: T): void {
    const key = this.generateKey(value);
    const element = this.elements.get(key);
    
    if (element) {
      // Remove all tags to delete the element
      element.tags = [];
      this.elements.delete(key);
    }
  }

  /**
   * Check if an element is in the set
   */
  has(value: T): boolean {
    const key = this.generateKey(value);
    const element = this.elements.get(key);
    
    if (!element) return false;
    return element.tags.length > 0;
  }

  /**
   * Get all elements in the set
   */
  values(): T[] {
    const result: T[] = [];
    
    for (const element of this.elements.values()) {
      if (element.tags.length > 0) {
        result.push(element.value);
      }
    }
    
    return result;
  }

  /**
   * Get the size of the set
   */
  size(): number {
    let count = 0;
    
    for (const element of this.elements.values()) {
      if (element.tags.length > 0) {
        count++;
      }
    }
    
    return count;
  }

  /**
   * Merge another OR-Set into this one
   */
  merge(other: ORSet<T>): void {
    for (const [key, otherElement] of other.elements.entries()) {
      const existing = this.elements.get(key);
      
      if (!existing) {
        // Element doesn't exist locally, copy it
        this.elements.set(key, {
          value: otherElement.value,
          tags: [...otherElement.tags],
        });
      } else {
        // Merge tags from both sets
        const mergedTags = this.mergeTags(existing.tags, otherElement.tags);
        
        if (mergedTags.length > 0) {
          this.elements.set(key, {
            value: existing.value,
            tags: mergedTags,
          });
        } else {
          // All tags were removed, delete the element
          this.elements.delete(key);
        }
      }
    }
  }

  /**
   * Merge tags from two sets, removing duplicates
   */
  private mergeTags(tags1: ORSetTag[], tags2: ORSetTag[]): ORSetTag[] {
    const tagMap = new Map<string, ORSetTag>();
    
    // Add all tags from first set
    for (const tag of tags1) {
      tagMap.set(tag.id, tag);
    }
    
    // Add or update tags from second set
    for (const tag of tags2) {
      const existing = tagMap.get(tag.id);
      
      if (existing) {
        // Merge vector clocks
        const mergedClock = VectorClockUtils.merge(
          existing.vectorClock,
          tag.vectorClock
        );
        tagMap.set(tag.id, {
          ...tag,
          vectorClock: mergedClock,
        });
      } else {
        tagMap.set(tag.id, tag);
      }
    }
    
    return Array.from(tagMap.values());
  }

  /**
   * Get the current state as a plain object
   */
  toJSON(): { elements: ORSetElement<T>[] } {
    return {
      elements: Array.from(this.elements.values()),
    };
  }

  /**
   * Create an OR-Set from a plain object
   */
  static fromJSON<T>(json: { elements: ORSetElement<T>[] }): ORSet<T> {
    const set = new ORSet<T>();
    
    for (const element of json.elements) {
      const key = JSON.stringify(element.value);
      set.elements.set(key, element);
    }
    
    return set;
  }

  /**
   * Clear all elements
   */
  clear(): void {
    this.elements.clear();
  }
}

File: server/src/crdt/or_set.rs
/**
 * Observed-Remove Set (OR-Set) CRDT
 * A set that handles concurrent adds and removes correctly
 */
use crate::crdt::vector_clock::VectorClock;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// A tag for an OR-Set element
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ORSetTag {
    pub id: String,
    pub node_id: String,
    pub vector_clock: VectorClock,
}

/// An element in an OR-Set
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ORSetElement<T> {
    pub value: T,
    pub tags: Vec<ORSetTag>,
}

/// Observed-Remove Set CRDT
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ORSet<T> {
    elements: HashMap<String, ORSetElement<T>>,
}

impl<T: Clone + PartialEq + serde::Serialize + for<'de> serde::Deserialize<'de>> ORSet<T> {
    /// Create a new OR-Set
    pub fn new() -> Self {
        Self {
            elements: HashMap::new(),
        }
    }

    /// Generate a unique tag for an element
    fn generate_tag(node_id: &str) -> ORSetTag {
        let id = format!(
            "{}-{}-{}",
            node_id,
            chrono::Utc::now().timestamp_millis(),
            uuid::Uuid::new_v4()
        );
        ORSetTag {
            id,
            node_id: node_id.to_string(),
            vector_clock: VectorClock::new(node_id, 1),
        }
    }

    /// Generate a unique key for an element value
    fn generate_key(value: &T) -> String {
        serde_json::to_string(value).unwrap_or_default()
    }

    /// Add an element to the set
    pub fn add(&mut self, value: T, node_id: &str) {
        let key = Self::generate_key(&value);
        let tag = Self::generate_tag(node_id);

        if let Some(element) = self.elements.get_mut(&key) {
            element.tags.push(tag);
        } else {
            self.elements.insert(
                key,
                ORSetElement {
                    value,
                    tags: vec![tag],
                },
            );
        }
    }

    /// Remove an element from the set
    pub fn remove(&mut self, value: &T) {
        let key = Self::generate_key(value);
        self.elements.remove(&key);
    }

    /// Check if an element is in the set
    pub fn has(&self, value: &T) -> bool {
        let key = Self::generate_key(value);
        if let Some(element) = self.elements.get(&key) {
            !element.tags.is_empty()
        } else {
            false
        }
    }

    /// Get all elements in the set
    pub fn values(&self) -> Vec<T> {
        let mut result = Vec::new();
        for element in self.elements.values() {
            if !element.tags.is_empty() {
                result.push(element.value.clone());
            }
        }
        result
    }

    /// Get the size of the set
    pub fn size(&self) -> usize {
        self.elements
            .values()
            .filter(|e| !e.tags.is_empty())
            .count()
    }

    /// Merge another OR-Set into this one
    pub fn merge(&mut self, other: &ORSet<T>) {
        for (key, other_element) in &other.elements {
            if let Some(existing) = self.elements.get_mut(key) {
                let merged_tags = Self::merge_tags(&existing.tags, &other_element.tags);
                if merged_tags.is_empty() {
                    self.elements.remove(key);
                } else {
                    existing.tags = merged_tags;
                }
            } else {
                self.elements.insert(key.clone(), other_element.clone());
            }
        }
    }

    /// Merge tags from two sets, removing duplicates
    fn merge_tags(tags1: &[ORSetTag], tags2: &[ORSetTag]) -> Vec<ORSetTag> {
        let mut tag_map: HashMap<String, ORSetTag> = HashMap::new();

        // Add all tags from first set
        for tag in tags1 {
            tag_map.insert(tag.id.clone(), tag.clone());
        }

        // Add or update tags from second set
        for tag in tags2 {
            if let Some(existing) = tag_map.get_mut(&tag.id) {
                // Merge vector clocks
                existing.vector_clock = existing.vector_clock.merge(&tag.vector_clock);
            } else {
                tag_map.insert(tag.id.clone(), tag.clone());
            }
        }

        tag_map.into_values().collect()
    }

    /// Clear all elements
    pub fn clear(&mut self) {
        self.elements.clear();
    }
}

impl<T: Clone + PartialEq + serde::Serialize + for<'de> serde::Deserialize<'de>> Default
    for ORSet<T>
{
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_or_set_add() {
        let mut set = ORSet::new();
        set.add("value1".to_string(), "node1");
        assert!(set.has(&"value1".to_string()));
    }

    #[test]
    fn test_or_set_remove() {
        let mut set = ORSet::new();
        set.add("value1".to_string(), "node1");
        set.remove(&"value1".to_string());
        assert!(!set.has(&"value1".to_string()));
    }

    #[test]
    fn test_or_set_merge() {
        let mut set1 = ORSet::new();
        set1.add("value1".to_string(), "node1");

        let mut set2 = ORSet::new();
        set2.add("value2".to_string(), "node2");

        set1.merge(&set2);
        assert!(set1.has(&"value1".to_string()));
        assert!(set1.has(&"value2".to_string()));
    }
}
